### PR TITLE
Data decoder

### LINKED ISSUE #121: Unmarshal/Decode in data package
It is inconveinent to process parameters passed as `data.Map` in plugins, especially when a plugin has many parameters. To help data conversions, type checking, finding missing required parameters, detecting undefined parameters, `data` package should provide something like mapstructure.

For example,

```
type SourceParams struct {
  RequiredParam1 Type `bql:",required"` // same as `bql:"required_param_1,required"`
  RequiredParam2 Type `bql:"custom_name,required"`
  OptionalParam1 Type // same as `bql:"optional_param_1"`
  OptionalParam2 Type `bql:"custom_name_2"`
}

p := SourceParams{
  OptionalParam1: defaultValue,
  OptionalParam2: defaultValue,
}

if err := data.Unmarshal(params, &p); err != nil {
  ...
}
```

Notes:
- Only `Unmarshal` or `Decode` will be provided in the first version
- `time.Duration` should be supported by using `data.ToDuration`
- Implicit type conversion should be supported by tag
- Default values should be given when creating an instance of a struct
- Something like `Metadata` in mapstructure should be provided to detect missing/unused keys
- Accepting ilke 1.0 as an integer may be required

### PR DESCRIPTION
**Pull request by nobu-k**
I implemented `data.Decoder` which decodes a struct from `data.Map`. This is intended to be used by plugins to define their parameters easily. Sorry for a pretty huge pull request.

fixes #121

### CODE REVIEW

File: data/decode.go [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+//	* struct, embedded struct
+//	* Blob
+//	* time.Time, Timestamp (see ToTimestamp to know supported values)
+//	* time.Duration (can be decoded from following types:)
+//		* int: second (e.g. 5 => 5 * time.Second)
+//		* float: second + subsecond (e.g 2.5 => 2 * time.Second + 500 * time.Millisecond)
+//		* string: Go's duration format (e.g. "6s" => 6 * time.Second)
+//	* pointer of these types
+//
+// A regular array is not supported yet. User defined time.Time-compatbile
```
Comment (disktnk) [inline data/decode.go]: -> "compatible"

File: data/decode.go [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+//
+// weaklytyped option is available to allow users to write values flexibly:
+//
+//	struct {
+//		Param1 int `bql:",weaklytyped"`
+//	}
+//
+// Internally, weaklytyped fields are converted by ToType function, while
+// AsType function is used for regular fields. Following types are always
+// cnosidered weaklytyped: time.Time, Timestamp, time.Duration. Struct and
```
Comment (disktnk) [inline data/decode.go]: -> "considered"

File: data/decode_test.go [hunk 3]
```diff
@@ -0,0 +1,708 @@
...
+				"f": Float(3.14),
+			}, s)
+
+			Convey("Then it should fail", func() {
+				fmt.Println(err)
+				So(err, ShouldNotBeNil)
+			})
+		})
+
+		Convey("When decoding non-addresable value", func() {
```
Comment (disktnk) [inline data/decode_test.go]: -> "addressable"

**Comment by disktnk** [comment]
Great Job!

memo: It has an error in TravisCI check, but the error is not related to this PR.

**Comment by nobu-k** [comment]
Thanks! I'll fix them on Thursday this week.

File: data/decode.go [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+		for ti, opt := range opts {
+			if ti == 0 { // skip name
+				continue
+			}
+			switch opt {
+			case "required":
+				required = true
+			case "weaklytyped":
+				weaklyTyped = true
+			default:
```
Comment (nobu-k) [inline data/decode.go]: I fixed it to return a special error message when the option is an empty string.

**Comment by nobu-k** [comment]
Thank you!

**Pull request merged**